File: src/sdk/go/subscribe.go
package appmesh

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// Sentinel errors returned by WaitForAsyncRun to disambiguate non-EXIT terminations
// from real (possibly negative) process exit codes (e.g. -SIGINT = -2).
var (
	ErrTransportDisconnected = errors.New("transport disconnected before exit")
	ErrAppRemoved            = errors.New("app removed before exit")
)

const (
	EVENT_URI = "/appmesh/event"

	// EventTypeDisconnected is a synthetic event_type pushed to every registered
	// callback when the demuxer stops or the underlying transport disconnects.
	// This lets long-running waits (e.g. wait_for_async_run) unblock instead of
	// hanging forever.
	EventTypeDisconnected = "__disconnected__"
)

// AppEvent represents a server-push event notification.
type AppEvent struct {
	SubscriptionID string          `json:"subscription_id"`
	EventType      string          `json:"event_type"`
	AppName        string          `json:"app_name"`
	Timestamp      int64           `json:"timestamp"`
	Sequence       uint64          `json:"sequence"`
	Data           json.RawMessage `json:"data"`
}

// EventCallback is invoked when a subscribed event arrives.
type EventCallback func(event AppEvent)

// MessageDemuxer reads messages from a transport connection and routes them
// to either pending request channels (by UUID) or event subscription callbacks.
// Event callbacks are dispatched serially via a single worker goroutine, so
// per-subscription event ordering is preserved (matching Python/Java SDKs).
type MessageDemuxer struct {
	mu       sync.Mutex
	pending  map[string]chan *Response // uuid -> response channel
	eventCBs map[string]EventCallback  // subscription_id -> callback
	// Events that arrive between server-side subscription and the client
	// registering its callback (e.g. atomic add_app(subscribe_events) on a fast
	// app, whose output is pushed before add_app returns). Held per subID and
	// flushed on registerEventCallback so no events are lost.
	eventBufs map[string][]AppEvent
	started   bool
	stopCh    chan struct{}
	eventQ    chan dispatchTask
	readMsg   func() ([]byte, error) // transport-specific read function
}

// Bound the pre-registration event buffer (atomic-subscribe race window) so a
// subscription whose callback never registers cannot grow memory without limit.
const (
	maxBufferedSubs         = 64
	maxBufferedEventsPerSub = 1000
)

type dispatchTask struct {
	cb    EventCallback
	event AppEvent
}

func newMessageDemuxer(readMsg func() ([]byte, error)) *MessageDemuxer {
	return &MessageDemuxer{
		pending:   make(map[string]chan *Response),
		eventCBs:  make(map[string]EventCallback),
		eventBufs: make(map[string][]AppEvent),
		stopCh:    make(chan struct{}),
		eventQ:    make(chan dispatchTask, 256),
		readMsg:   readMsg,
	}
}

// start begins the background reader and dispatch goroutines.
func (d *MessageDemuxer) start() {
	d.mu.Lock()
	if d.started {
		d.mu.Unlock()
		return
	}
	d.started = true
	d.mu.Unlock()

	go d.readLoop()
	go d.dispatchWorker()
}

// dispatchWorker invokes event callbacks serially in arrival order.
// One slow callback delays subsequent events but never the socket reader.
func (d *MessageDemuxer) dispatchWorker() {
	for {
		select {
		case <-d.stopCh:
			return
		case task := <-d.eventQ:
			func() {
				defer func() {
					if r := recover(); r != nil {
						log.Printf("dispatchWorker: callback panic: %v", r)
					}
				}()
				task.cb(task.event)
			}()
		}
	}
}

// stop terminates the background reader.
func (d *MessageDemuxer) stop() {
	d.mu.Lock()
	if !d.started {
		d.mu.Unlock()
		return
	}
	close(d.stopCh)
	d.started = false

	// Snapshot callbacks + pending under lock, then release before invoking
	type cbEntry struct {
		subID string
		cb    EventCallback
	}
	snapshot := make([]cbEntry, 0, len(d.eventCBs))
	for subID, cb := range d.eventCBs {
		snapshot = append(snapshot, cbEntry{subID, cb})
	}
	pendingCopy := d.pending
	d.pending = make(map[string]chan *Response)
	d.eventBufs = make(map[string][]AppEvent) // drop events buffered for never-registered subs
	d.mu.Unlock()

	// Close pending channels (outside lock)
	for _, ch := range pendingCopy {
		close(ch)
	}

	// Broadcast disconnect (outside lock — no deadlock from re-entrant callbacks)
	for _, e := range snapshot {
		func() {
			defer func() {
				if r := recover(); r != nil {
					log.Printf("broadcastDisconnect: callback panic: %v", r)
				}
			}()
			e.cb(AppEvent{SubscriptionID: e.subID, EventType: EventTypeDisconnected})
		}()
	}
}

// registerRequest creates a channel for a pending request identified by UUID.
func (d *MessageDemuxer) registerRequest(uuid string) chan *Response {
	ch := make(chan *Response, 1)
	d.mu.Lock()
	d.pending[uuid] = ch
	d.mu.Unlock()
	return ch
}

// unregisterRequest removes a pending request channel.
func (d *MessageDemuxer) unregisterRequest(uuid string) {
	d.mu.Lock()
	delete(d.pending, uuid)
	d.mu.Unlock()
}

// registerEventCallback registers a callback for a subscription ID, flushing any
// events that arrived before registration (atomic-subscribe race).
func (d *MessageDemuxer) registerEventCallback(subID string, cb EventCallback) {
	d.mu.Lock()
	d.eventCBs[subID] = cb
	buffered := d.eventBufs[subID]
	delete(d.eventBufs, subID)
	// Enqueue under the lock so buffered events precede later live events.
	for _, event := range buffered {
		select {
		case d.eventQ <- dispatchTask{cb: cb, event: event}:
		case <-d.stopCh:
		}
	}
	d.mu.Unlock()
}

// unregisterEventCallback removes a callback for a subscription ID.
func (d *MessageDemuxer) unregisterEventCallback(subID string) {
	d.mu.Lock()
	delete(d.eventCBs, subID)
	delete(d.eventBufs, subID)
	d.mu.Unlock()
}

// readLoop continuously reads messages and routes them.
func (d *MessageDemuxer) readLoop() {
	for {
		select {
		case <-d.stopCh:
			return
		default:
		}

		data, err := d.readMsg()
		if err != nil {
			// Connection error — stop() handles broadcast + cleanup
			d.stop()
			return
		}

		resp := &Response{}
		if err := resp.Deserialize(data); err != nil {
			continue
		}

		if resp.RequestUri == EVENT_URI {
			d.dispatchEvent(resp)
		} else {
			d.dispatchResponse(resp)
		}
	}
}

// dispatchEvent routes an event push to the matching subscription callback.
func (d *MessageDemuxer) dispatchEvent(resp *Response) {
	var event AppEvent
	if err := json.Unmarshal(resp.Body, &event); err != nil {
		return
	}

	subID := event.SubscriptionID
	if subID == "" {
		if v, ok := resp.Headers["X-Subscription-Id"]; ok {
			subID = v
		}
	}

	d.mu.Lock()
	cb, ok := d.eventCBs[subID]
	if (!ok || cb == nil) && subID != "" {
		// No callback yet: buffer the event so an atomic add_app(subscribe_events)
		// on a fast app does not lose output pushed before the client registers.
		d.bufferEventLocked(subID, event)
		d.mu.Unlock()
		return
	}
	d.mu.Unlock()

	if !ok || cb == nil {
		return
	}
	// Enqueue for serial dispatch. Order matters — STDOUT dedup by byte-position
	// in WaitForAsyncRun assumes events arrive in the order the daemon emitted them.
	select {
	case d.eventQ <- dispatchTask{cb: cb, event: event}:
	case <-d.stopCh:
	}
}

// bufferEventLocked holds an event whose callback has not registered yet.
// The caller must hold d.mu. Bounded: caps distinct buffered subIDs and uses
// drop-oldest per subID so a never-registered subscription cannot grow memory
// without limit.
func (d *MessageDemuxer) bufferEventLocked(subID string, event AppEvent) {
	buf, exists := d.eventBufs[subID]
	if !exists && len(d.eventBufs) >= maxBufferedSubs {
		return // cap distinct unregistered subs to bound memory
	}
	if len(buf) >= maxBufferedEventsPerSub {
		buf = buf[1:] // drop oldest
	}
	d.eventBufs[subID] = append(buf, event)
}

// dispatchResponse routes a request response to the matching pending channel.
func (d *MessageDemuxer) dispatchResponse(resp *Response) {
	d.mu.Lock()
	ch, ok := d.pending[resp.UUID]
	if ok {
		delete(d.pending, resp.UUID)
	}
	d.mu.Unlock()

	if ok && ch != nil {
		ch <- resp
	}
}

// SubscribeOption configures a subscription request.
type SubscribeOption struct {
	AppName string   // App name, or empty/"*" for all apps
	Events  []string // Event types: "START", "EXIT", "STDOUT", "HEALTH", "STATUS", "REMOVED"
}

// SubscriptionResult contains the server's response to a subscribe request.
type SubscriptionResult struct {
	SubscriptionID string   `json:"subscription_id"`
	AppName        string   `json:"app_name"`
	Events         []string `json:"events"`
}

// Subscribe registers for events on a named app (or all apps if AppName is empty/"*").
// The callback is invoked in a separate goroutine for each event.
// Returns the subscription ID and any error.
func (c *AppMeshClient) Subscribe(opt SubscribeOption, callback EventCallback) (*SubscriptionResult, error) {
	apiPath := "/appmesh/subscribe"
	if opt.AppName != "" && opt.AppName != "*" {
		apiPath = fmt.Sprintf("/appmesh/app/%s/subscribe", opt.AppName)
	}

	queries := url.Values{}
	if len(opt.Events) > 0 {
		queries.Set("events", strings.Join(opt.Events, ","))
	}

	status, raw, _, err := c.req.Send(http.MethodPost, apiPath, queries, nil, nil)
	if err != nil {
		return nil, err
	}
	if status != http.StatusOK {
		return nil, fmt.Errorf("subscribe failed with status %d: %s", status, string(raw))
	}

	var result SubscriptionResult
	if err := json.Unmarshal(raw, &result); err != nil {
		return nil, fmt.Errorf("failed to parse subscribe response: %w", err)
	}
	if result.SubscriptionID == "" {
		return nil, fmt.Errorf("server returned empty subscription_id")
	}

	// Register the callback with the demuxer (if transport supports it).
	if sub, ok := c.req.(subscribableRequester); ok {
		sub.enableDemuxer()
		if d := sub.getDemuxer(); d != nil {
			d.registerEventCallback(result.SubscriptionID, callback)
		}
	}

	return &result, nil
}

// Unsubscribe removes a subscription by ID.
func (c *AppMeshClient) Unsubscribe(subscriptionID string) error {
	queries := url.Values{}
	queries.Set("subscription_id", subscriptionID)

	status, raw, _, err := c.req.Send(http.MethodDelete, "/appmesh/subscribe", queries, nil, nil)
	if err != nil {
		return err
	}
	if status != http.StatusOK && status != http.StatusNotFound {
		return fmt.Errorf("unsubscribe failed with status %d: %s", status, string(raw))
	}

	// Unregister from demuxer (no-op if transport is closed or demuxer not enabled).
	if sub, ok := c.req.(subscribableRequester); ok {
		if d := sub.getDemuxer(); d != nil {
			d.unregisterEventCallback(subscriptionID)
		}
	}

	return nil
}

// WaitForAsyncRun waits for an asynchronous application run to complete using
// subscribe-based streaming instead of polling. It subscribes to STDOUT, EXIT,
// and REMOVED events, then backfills any output emitted before the subscription
// took effect. Stdout is deduplicated by byte position to handle overlap between
// the backfill and live events.
//
// Return values:
//   - (&code, nil):                  process exited normally (code may be negative for signal kills)
//   - (nil, nil):                    timeout or context cancelled
//   - (nil, ErrAppRemoved):          app removed before EXIT observed
//   - (nil, ErrTransportDisconnected): TCP/WSS connection lost
//   - (nil, err):                    subscribe failed
func (c *AppMeshClient) WaitForAsyncRun(ctx context.Context, run *AppRun, stdoutHandler OutputHandler, timeout time.Duration) (*int, error) {
	if run == nil || run.AppName == "" {
		return nil, fmt.Errorf("invalid async run")
	}

	var (
		exitCode       *int       // set on EXIT
		waitErr        error      // first non-EXIT termination reason
		disconnected   bool       // transport-dead flag (set even if waitErr is already populated)
		deliveredUntil int64      // next-byte offset already delivered to handler
		mu             sync.Mutex // guards all of the above
		done           = make(chan struct{})
		doneOnce       sync.Once
	)

	signalDone := func() { doneOnce.Do(func() { close(done) }) }

	setExit := func(code int) {
		mu.Lock()
		if exitCode == nil && waitErr == nil {
			exitCode = &code
		}
		mu.Unlock()
		signalDone()
	}

	setErr := func(e error) {
		mu.Lock()
		if exitCode == nil && waitErr == nil {
			waitErr = e
		}
		mu.Unlock()
		signalDone()
	}

	// deliver passes the not-yet-seen portion of chunk to the handler,
	// using byte-position dedup to bridge backfill ↔ live events.
	deliver := func(chunk string, pos int64) {
		if len(chunk) == 0 {
			return
		}
		mu.Lock()
		end := pos + int64(len(chunk))
		if end <= deliveredUntil {
			mu.Unlock()
			return
		}
		startPos := pos
		if pos < deliveredUntil {
			chunk = chunk[deliveredUntil-pos:]
			startPos = deliveredUntil
		}
		deliveredUntil = end
		mu.Unlock()

		if stdoutHandler != nil {
			stdoutHandler(chunk, startPos)
		}
	}

	onEvent := func(event AppEvent) {
		switch event.EventType {
		case "STDOUT":
			var data struct {
				Output   string `json:"output"`
				Position int64  `json:"position"`
			}
			if err := json.Unmarshal(event.Data, &data); err != nil {
				return
			}
			deliver(data.Output, data.Position)

		case "EXIT":
			var data struct {
				ExitCode *int `json:"exit_code"`
			}
			if err := json.Unmarshal(event.Data, &data); err == nil && data.ExitCode != nil {
				setExit(*data.ExitCode)
			} else {
				setErr(fmt.Errorf("EXIT event missing exit_code"))
			}

		case "REMOVED":
			setErr(ErrAppRemoved)

		case EventTypeDisconnected:
			// Always mark disconnected (even if waitErr already set) so the
			// defer below skips Unsubscribe on a dead connection.
			mu.Lock()
			disconnected = true
			switch {
			case exitCode != nil:
				// Exit already reported — leave it alone, the process finished
				// successfully (or with a normal code) before the transport died.
			case waitErr == nil:
				waitErr = ErrTransportDisconnected
			default:
				// Wrap the existing reason so callers using errors.Is can still
				// detect ErrTransportDisconnected (race: REMOVED/malformed-EXIT
				// arrived just before the transport dropped).
				waitErr = fmt.Errorf("%w (also: %w)", waitErr, ErrTransportDisconnected)
			}
			mu.Unlock()
			signalDone()
		}
	}

	sub, err := c.Subscribe(SubscribeOption{
		AppName: run.AppName,
		Events:  []string{"STDOUT", "EXIT", "REMOVED"},
	}, onEvent)
	if err != nil {
		log.Printf("WaitForAsyncRun: subscribe failed for %s: %v", run.AppName, err)
		return nil, fmt.Errorf("subscribe: %w", err)
	}

	defer func() {
		mu.Lock()
		disc := disconnected
		mu.Unlock()
		// On disconnect the demuxer is stopped; calling Unsubscribe would
		// register a request channel that never gets a response.
		if disc {
			return
		}
		if sub.SubscriptionID != "" {
			_ = c.Unsubscribe(sub.SubscriptionID)
		}
	}()

	// Backfill: fetch output emitted before the subscribe took effect, and
	// detect a process that already exited.
	backfill := c.GetAppOutput(run.AppName, 0, 0, 0, run.ProcUid, 0)
	if backfill.HttpBody != "" {
		deliver(backfill.HttpBody, 0)
	}
	if backfill.ExitCode != nil {
		setExit(*backfill.ExitCode)
	}

	// Wait for completion, timeout, or context cancellation.
	if timeout <= 0 {
		select {
		case <-done:
		case <-ctx.Done():
		}
	} else {
		select {
		case <-done:
		case <-time.After(timeout):
		case <-ctx.Done():
		}
	}

	mu.Lock()
	defer mu.Unlock()
	return exitCode, waitErr
}

// EnableConcurrency makes this client's transport safe for concurrent use by multiple
// goroutines. It starts the response demuxer, which correlates each reply to its request
// by UUID; without it a TCP/WSS connection shared across goroutines uses a synchronous
// send-then-read that can cross-wire responses (e.g. a token-renew reply consumed by an
// unrelated call, leaving a stale/revoked token in place). No-op on transports without
// multiplexing (HTTP) and idempotent (Subscribe also enables it on demand).
func (c *AppMeshClient) EnableConcurrency() {
	if sub, ok := c.req.(subscribableRequester); ok {
		sub.enableDemuxer()
	}
}

// subscribableRequester is implemented by transports that support event subscriptions.
type subscribableRequester interface {
	enableDemuxer()
	getDemuxer() *MessageDemuxer
}

File: src/sdk/rust/src/subscribe.rs
// subscribe.rs
//! Message demuxer for routing TCP/WSS responses and event notifications.
//!
//! When a subscription is active the background reader owns the transport's
//! read side.  Incoming messages are routed by inspecting `request_uri`:
//!   - `/appmesh/event` -> dispatched to the registered event callback
//!   - anything else    -> dispatched to the pending oneshot channel by UUID

use std::collections::{HashMap, VecDeque};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

use async_trait::async_trait;
use log::{debug, error, warn};
use tokio::sync::{mpsc, oneshot, watch, Mutex};

use crate::constants::{EVENT_TYPE_DISCONNECTED, EVENT_URI};
use crate::error::AppMeshError;
use crate::models::AppEvent;
use crate::wire_messages::ResponseMessage;

/// Callback invoked for each received event. Events for a subscription are
/// delivered serially in arrival order on a dedicated dispatch task (matching the
/// other SDK demuxers), so a slow callback never blocks the reader loop or other subs.
pub type EventCallback = Arc<dyn Fn(AppEvent) + Send + Sync>;

/// Trait abstracting the transport read side so the demuxer works with both
/// TCP and WSS transports.
#[async_trait]
pub(crate) trait MessageReader: Send + Sync {
    async fn read_message(&self) -> Result<Option<Vec<u8>>, AppMeshError>;
}

/// Bound the pre-registration event buffer (atomic-subscribe race window) so a
/// subscription whose callback never registers cannot grow memory without limit.
const MAX_BUFFERED_SUBS: usize = 64;
const MAX_BUFFERED_EVENTS_PER_SUB: usize = 1000;

/// State protected together by a single mutex so the dispatch-channel map and
/// the pre-registration event buffer stay consistent (events buffered while a
/// callback is absent are flushed atomically on registration, preserving order
/// vs. concurrent live events).
struct EventState {
    /// Per-subscription dispatch channels: one worker task per sub_id drains its
    /// channel and invokes the callback, so buffered flushes, live events, and the
    /// disconnect broadcast are delivered serially in arrival order.
    senders: HashMap<String, mpsc::UnboundedSender<AppEvent>>,
    /// Events that arrive between the server-side subscription and the client
    /// registering its callback (e.g. atomic add_app(subscribe_events) on a fast
    /// app, whose output is pushed before the response returns). Held per sub_id
    /// and flushed on register_event_callback so no events are lost.
    buffers: HashMap<String, VecDeque<AppEvent>>,
}

/// Routes incoming messages to pending request channels or event callbacks.
pub struct MessageDemuxer {
    pending: Arc<Mutex<HashMap<String, oneshot::Sender<ResponseMessage>>>>,
    event_state: Arc<std::sync::Mutex<EventState>>,
    stop_tx: std::sync::Mutex<Option<watch::Sender<bool>>>,
    running: Arc<AtomicBool>,
}

impl MessageDemuxer {
    /// Create a new (stopped) demuxer.
    pub fn new() -> Self {
        Self {
            pending: Arc::new(Mutex::new(HashMap::new())),
            event_state: Arc::new(std::sync::Mutex::new(EventState {
                senders: HashMap::new(),
                buffers: HashMap::new(),
            })),
            stop_tx: std::sync::Mutex::new(None),
            running: Arc::new(AtomicBool::new(false)),
        }
    }

    /// Start the background read loop.  No-op if already running.
    pub(crate) fn start(&self, reader: Arc<dyn MessageReader>) {
        if self.running.swap(true, Ordering::SeqCst) {
            return; // already running
        }

        let (stop_tx, stop_rx) = watch::channel(false);
        // Poisoning is benign here (guarded state stays valid), so recover the guard.
        *self.stop_tx.lock().unwrap_or_else(|e| e.into_inner()) = Some(stop_tx);

        let pending = Arc::clone(&self.pending);
        let event_state = Arc::clone(&self.event_state);
        let running = Arc::clone(&self.running);

        tokio::spawn(async move {
            Self::read_loop(reader, stop_rx, pending, event_state, running).await;
        });
    }

    /// Stop the background reader and fail all pending requests.
    pub fn stop(&self) {
        if !self.running.swap(false, Ordering::SeqCst) {
            return; // already stopped
        }

        // Broadcast a synthetic disconnect event to all registered event
        // callbacks so long-running waits can unblock cleanly, and discard any
        // events buffered for never-registered subs so nothing leaks.
        Self::broadcast_disconnect(&self.event_state);

        // Signal the read loop to exit
        if let Some(tx) = self.stop_tx.lock().unwrap_or_else(|e| e.into_inner()).take() {
            let _ = tx.send(true);
        }

        // Drop all pending senders so waiters get a RecvError. stop() may run
        // outside a tokio runtime (e.g. from Drop): spawn when possible, else
        // clear inline — with no runtime nothing holds the async lock.
        let pending = self.pending.clone();
        if let Ok(handle) = tokio::runtime::Handle::try_current() {
            handle.spawn(async move {
                let mut map = pending.lock().await;
                map.clear();
            });
        } else if let Ok(mut map) = pending.try_lock() {
            map.clear();
        }
    }

    /// Register a pending request and return the receiver end.
    ///
    /// The caller should `await` the receiver; the read loop will send the
    /// matching `ResponseMessage` when it arrives.
    pub async fn register_request(&self, uuid: &str) -> oneshot::Receiver<ResponseMessage> {
        let (tx, rx) = oneshot::channel();
        let mut map = self.pending.lock().await;
        // Re-check `running` under the pending lock: stop()/read-loop exit flip it
        // false BEFORE clearing the map, so any clear after a true reading must wait
        // on this lock and will drop our sender (waking the waiter). Inserting after
        // the clear would strand the sender forever and hang the waiter.
        if self.running.load(Ordering::SeqCst) {
            map.insert(uuid.to_string(), tx);
        }
        // When stopped, `tx` drops here so `rx.await` fails immediately instead of hanging.
        rx
    }

    /// Remove a pending request (e.g. on timeout / cancellation).
    pub async fn unregister_request(&self, uuid: &str) {
        let mut map = self.pending.lock().await;
        map.remove(uuid);
    }

    /// Register an event callback for a subscription id, flushing any events
    /// that arrived before registration (atomic-subscribe race).
    pub fn register_event_callback(&self, sub_id: &str, callback: EventCallback) {
        let (tx, mut rx) = mpsc::unbounded_channel::<AppEvent>();

        // Single dispatch worker per subscription: drains the channel and invokes
        // the callback serially in arrival order.
        tokio::spawn(async move {
            while let Some(event) = rx.recv().await {
                // Isolate callback panics (e.g. print! on a closed stdout pipe) so
                // later events — including EXIT and DISCONNECTED — are still delivered.
                let call = std::panic::AssertUnwindSafe(|| callback(event));
                if std::panic::catch_unwind(call).is_err() {
                    error!("event callback panicked; continuing event dispatch");
                }
            }
        });

        // Flush buffered events and publish the sender under the same lock so live
        // events can only be enqueued after the buffered ones (order preserved).
        let mut state = self.event_state.lock().unwrap_or_else(|e| e.into_inner());
        if let Some(buffered) = state.buffers.remove(sub_id) {
            for event in buffered {
                let _ = tx.send(event);
            }
        }
        // Replacing an existing sender drops it, stopping the old worker after it drains.
        state.senders.insert(sub_id.to_string(), tx);
    }

    /// Remove an event callback and discard any events buffered for it.
    pub fn unregister_event_callback(&self, sub_id: &str) {
        let mut state = self.event_state.lock().unwrap_or_else(|e| e.into_inner());
        // Dropping the sender stops the dispatch worker after it drains.
        state.senders.remove(sub_id);
        state.buffers.remove(sub_id);
    }

    /// Whether the background reader is currently running.
    pub fn is_running(&self) -> bool {
        self.running.load(Ordering::SeqCst)
    }

    // -- internal -------------------------------------------------------------

    /// Push a synthetic disconnect event to every registered event callback,
    /// and discard any events buffered for never-registered subs.
    fn broadcast_disconnect(event_state: &Arc<std::sync::Mutex<EventState>>) {
        let mut state = event_state.lock().unwrap_or_else(|e| e.into_inner());
        state.buffers.clear();

        // Enqueue through the per-subscription dispatch channels so the
        // disconnect event is observed after any events already in flight.
        for (sub_id, sender) in state.senders.iter() {
            let event = AppEvent {
                subscription_id: sub_id.clone(),
                event_type: EVENT_TYPE_DISCONNECTED.to_string(),
                app_name: String::new(),
                timestamp: 0,
                sequence: 0,
                data: serde_json::Value::Null,
            };
            let _ = sender.send(event);
        }
    }

    async fn read_loop(
        reader: Arc<dyn MessageReader>,
        mut stop_rx: watch::Receiver<bool>,
        pending: Arc<Mutex<HashMap<String, oneshot::Sender<ResponseMessage>>>>,
        event_state: Arc<std::sync::Mutex<EventState>>,
        running: Arc<AtomicBool>,
    ) {
        debug!("MessageDemuxer: read loop started");

        loop {
            // Check for stop signal (non-blocking)
            if *stop_rx.borrow() {
                break;
            }

            let result = tokio::select! {
                r = reader.read_message() => r,
                _ = stop_rx.changed() => break,
            };

            match result {
                Ok(Some(data)) if !data.is_empty() => {
                    match ResponseMessage::deserialize(&data) {
                        Ok(resp) => {
                            if resp.request_uri == EVENT_URI {
                                Self::dispatch_event(&resp, &event_state);
                            } else {
                                Self::dispatch_response(resp, &pending).await;
                            }
                        }
                        Err(e) => {
                            warn!("MessageDemuxer: failed to deserialize message: {}", e);
                        }
                    }
                }
                Ok(_) => {
                    // Empty data or None means connection closed
                    debug!("MessageDemuxer: connection closed");
                    break;
                }
                Err(e) => {
                    error!("MessageDemuxer: read error: {}", e);
                    break;
                }
            }
        }

        // Only broadcast if stop() hasn't already done so
        if running.swap(false, Ordering::SeqCst) {
            Self::broadcast_disconnect(&event_state);
        }

        let mut map = pending.lock().await;
        map.clear();

        debug!("MessageDemuxer: read loop exited");
    }

    fn dispatch_event(resp: &ResponseMessage, event_state: &Arc<std::sync::Mutex<EventState>>) {
        let event: AppEvent = match serde_json::from_slice(&resp.body) {
            Ok(e) => e,
            Err(e) => {
                warn!("MessageDemuxer: failed to parse event: {}", e);
                return;
            }
        };

        let mut sub_id = event.subscription_id.clone();
        if sub_id.is_empty() {
            if let Some(id) = resp.headers.get("X-Subscription-Id") {
                sub_id = id.clone();
            }
        }

        let mut state = event_state.lock().unwrap_or_else(|e| e.into_inner());
        if let Some(sender) = state.senders.get(&sub_id) {
            debug!("MessageDemuxer: event matched sub_id={}", sub_id);
            // Unbounded send never blocks; the dispatch worker preserves order.
            let _ = sender.send(event);
        } else if !sub_id.is_empty() {
            // Race: event arrived before its callback registered (e.g. atomic
            // add_app(subscribe_events) on a fast app). Buffer it (bounded) so
            // register_event_callback can flush it instead of dropping it.
            Self::buffer_event(&mut state, &sub_id, event);
        } else {
            warn!("MessageDemuxer: event dropped, empty sub_id");
        }
    }

    /// Hold an event whose callback has not registered yet (caller holds the lock).
    fn buffer_event(state: &mut EventState, sub_id: &str, event: AppEvent) {
        let buf = match state.buffers.get_mut(sub_id) {
            Some(buf) => buf,
            None => {
                if state.buffers.len() >= MAX_BUFFERED_SUBS {
                    debug!("MessageDemuxer: event sub_id={} dropped, buffer cap reached", sub_id);
                    return; // cap distinct unregistered subs to bound memory
                }
                state.buffers.entry(sub_id.to_string()).or_default()
            }
        };
        if buf.len() >= MAX_BUFFERED_EVENTS_PER_SUB {
            buf.pop_front(); // drop-oldest
        }
        buf.push_back(event);
    }

    async fn dispatch_response(
        resp: ResponseMessage,
        pending: &Arc<Mutex<HashMap<String, oneshot::Sender<ResponseMessage>>>>,
    ) {
        let uuid = resp.uuid.clone();
        let tx = {
            let mut map = pending.lock().await;
            map.remove(&uuid)
        };

        if let Some(sender) = tx {
            if sender.send(resp).is_err() {
                debug!("MessageDemuxer: receiver dropped for uuid {}", uuid);
            }
        } else {
            debug!("MessageDemuxer: no pending request for uuid {}", uuid);
        }
    }
}

impl Drop for MessageDemuxer {
    fn drop(&mut self) {
        // Best-effort stop without spawning
        self.running.store(false, Ordering::SeqCst);
        if let Some(tx) = self.stop_tx.lock().unwrap_or_else(|e| e.into_inner()).take() {
            let _ = tx.send(true);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::time::Duration;

    /// Scripted mock transport: messages are fed through a channel; closing the
    /// channel reads as transport EOF.
    struct ScriptedReader {
        rx: Mutex<mpsc::UnboundedReceiver<Vec<u8>>>,
    }

    #[async_trait]
    impl MessageReader for ScriptedReader {
        async fn read_message(&self) -> Result<Option<Vec<u8>>, AppMeshError> {
            Ok(self.rx.lock().await.recv().await)
        }
    }

    fn response_bytes(uuid: &str, request_uri: &str, body: &[u8]) -> Vec<u8> {
        let resp = ResponseMessage {
            uuid: uuid.to_string(),
            request_uri: request_uri.to_string(),
            http_status: 200,
            body: body.to_vec(),
            ..Default::default()
        };
        // Struct-map encoding, matching the daemon wire format
        // (see RequestMessage::serialize).
        rmp_serde::to_vec_named(&resp).expect("serialize response")
    }

    // Conformance: S7 (partial) — pending waiter registered before the response
    // arrives is routed by UUID; see docs/source/SDKContract.md.
    #[tokio::test]
    async fn conformance_s7_response_routed_to_pre_registered_waiter() {
        let (tx, rx) = mpsc::unbounded_channel();
        let demuxer = MessageDemuxer::new();
        demuxer.start(Arc::new(ScriptedReader { rx: Mutex::new(rx) }));

        let waiter = demuxer.register_request("req-s7").await;
        tx.send(response_bytes("req-s7", "/appmesh/app/test", b"{}")).unwrap();

        let resp = tokio::time::timeout(Duration::from_secs(2), waiter)
            .await
            .expect("timed out waiting for response")
            .expect("waiter channel failed");
        assert_eq!(resp.uuid, "req-s7");
        assert_eq!(resp.http_status, 200);
        demuxer.stop();
    }

    // Conformance: S2 (demuxer) — transport EOF broadcasts the synthetic
    // __disconnected__ event to every registered callback and fails pending
    // request waiters instead of leaving them hanging; see
    // docs/source/SDKContract.md.
    #[tokio::test]
    async fn conformance_s2_disconnect_broadcast_unblocks() {
        let (tx, rx) = mpsc::unbounded_channel::<Vec<u8>>();
        let demuxer = MessageDemuxer::new();
        demuxer.start(Arc::new(ScriptedReader { rx: Mutex::new(rx) }));

        let (event_tx, mut event_rx) = mpsc::unbounded_channel();
        demuxer.register_event_callback(
            "sub-s2",
            Arc::new(move |event: AppEvent| {
                let _ = event_tx.send(event);
            }),
        );
        let waiter = demuxer.register_request("req-s2").await;

        drop(tx); // transport EOF

        let event = tokio::time::timeout(Duration::from_secs(2), event_rx.recv())
            .await
            .expect("timed out waiting for disconnect broadcast")
            .expect("dispatch worker dropped");
        assert_eq!(event.event_type, EVENT_TYPE_DISCONNECTED);
        assert_eq!(event.subscription_id, "sub-s2");

        // Pending waiter is woken with an error (disconnect), never "slow request".
        let result = tokio::time::timeout(Duration::from_secs(2), waiter)
            .await
            .expect("pending waiter not woken on disconnect");
        assert!(result.is_err());
    }
}
